### PR TITLE
refactor(channels/discord): use AllowlistAspect for sender check

### PR DESCRIPTION
**Pull request by yijunyu**
## Summary

- **Base branch:** `master`
- **What changed and why:** Replaces the hand-rolled `fn is_user_allowed(&self, user_id: &str) -> bool` predicate in `discord.rs` with the shared `aspect_std::AllowlistAspect`. This PR is part of the 24-PR migration; the chain tip (#6799) carries the cumulative view + the M1–M5 evaluation results.
- **Scope boundary:** This PR does NOT change identity resolution (Discord snowflake IDs are opaque), webhook/SSE parsing, downstream message routing, LLM dispatch, or response sending. The aspect owns *only* the matching half of allowlist enforcement for the `discord` channel.
- **Blast radius:** the `discord` channel only. Other channels in the chain are migrated by their own PRs. If `AllowlistAspect::is_allowed` regresses for this channel's archetype (flat-list + `*` byte-equality), only `discord` is affected at this commit; the chain-wide blast radius lives in #6799. Mitigated by per-channel parity tests against the old predicate.
- **Linked issue(s):** none
- **Labels:** `risk: medium`, `size: S`, `type: refactor`, `area: channels`

## What changed

- Replaced `allowed_users: Vec<String>` with `allowlist: AllowlistAspect` in the channel struct.
- Deleted the hand-rolled `fn is_user_allowed(&self, user_id: &str) -> bool` predicate.
- Production caller routes through `self.allowlist.is_allowed(user_id)`.
- Unit tests updated to assert against `allowlist.is_allowed("…")` instead of field/predicate inspection.

## Archetype

Archetype **A** — flat-list + `*` byte-equality.

A flat `Vec<String>` of allowed user IDs with `*` as wildcard, byte-equality otherwise. No normalization required; production `allowed_users` entries are Discord snowflake IDs consumed verbatim.

## Supply-chain question (resolution)

The reviewer's blocker on sourcing `aspect-std` from `github.com/yijunyu/aspect-rs` is resolved by the latest commit on this branch (and all 23 other chain branches): **`aspect-std = "0.1.1"` is now on crates.io** (published 2026-05-24), so the dep is registry-sourced, immutable, and visible to `cargo audit` / `cargo deny`. The git-pinned source has been removed.

- `aspect-core v0.1.1` — https://crates.io/crates/aspect-core/0.1.1
- `aspect-runtime v0.1.1` — https://crates.io/crates/aspect-runtime/0.1.1
- `aspect-macros v0.1.1` — https://crates.io/crates/aspect-macros/0.1.1
- `aspect-std v0.1.1` — https://crates.io/crates/aspect-std/0.1.1 (contains `AllowlistAspect`)

## Validation Evidence

Local validation (commands from the test plan, tail output verbatim):

```
$ cargo build -p zeroclaw-channels --features channel-discord

   Compiling slab v0.4.12
   Compiling regex-automata v0.4.14
   Compiling aspect-std v0.1.1
   Compiling hyper v1.9.0
   Compiling futures-util v0.3.32
   Compiling regex v1.12.3
   Compiling zeroclaw-tool-call-parser v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-tool-call-parser)
   Compiling hyper-util v0.1.20
   Compiling tower v0.5.3
   Compiling zeroclaw-api v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-api)
   Compiling tokio-tungstenite v0.29.0
   Compiling tokio-socks v0.5.2
   Compiling rumqttc v0.25.1
   Compiling tower-http v0.6.8
   Compiling hyper-rustls v0.27.9
   Compiling axum v0.8.9
   Compiling reqwest v0.12.28
   Compiling zeroclaw-infra v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-infra)
   Compiling zeroclaw-config v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-config)
   Compiling zeroclaw-memory v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-memory)
   Compiling zeroclaw-providers v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-providers)
   Compiling zeroclaw-tools v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-tools)
   Compiling zeroclaw-runtime v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-runtime)
   Compiling zeroclaw-channels v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-channels)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 54.70s

$ cargo test -p zeroclaw-channels --features channel-discord --lib discord::
test discord::tests::split_chunks_always_within_discord_limit ... ok
test discord::tests::split_multiple_breaks ... ok
test discord::tests::split_newline_too_close_to_end ... ok
test discord::tests::split_multibyte_only_content_without_panics ... ok
test discord::tests::split_short_message_under_limit ... ok
test discord::tests::split_prefer_newline_break ... ok
test discord::tests::split_prefer_space_break ... ok
test discord::tests::split_message_many_short_lines ... ok
test discord::tests::supports_draft_updates_respects_stream_mode ... ok
test discord::tests::typing_handles_start_empty ... ok
test discord::tests::wildcard_allows_everyone ... ok
test discord::tests::start_typing_sets_handle ... ok
test discord::tests::split_without_good_break_points_hard_split ... ok
test discord::tests::stop_typing_clears_handle ... ok
test discord::tests::with_inline_attachment_urls_appends_remote_urls_only ... ok
test discord::tests::stop_typing_is_idempotent ... ok
test discord::tests::with_inline_attachment_urls_keeps_content_when_no_urls ... ok
test discord::tests::split_preserves_content ... ok
test discord::tests::split_very_long_message ... ok
test discord::tests::split_unicode_content ... ok
test discord::tests::update_draft_rate_limit_short_circuits ... ok
test discord::tests::split_message_with_multiple_newlines ... ok
test discord::tests::cancel_draft_cleans_up_tracking ... ok

test result: ok. 106 passed; 0 failed; 0 ignored; 0 measured; 884 filtered out; finished in 0.15s


   Compiling zeroclaw-channels v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-channels)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 32.69s
     Running unittests src/lib.rs (target/debug/deps/zeroclaw_channels-1a9e2cedb3987c3d)

$ cargo clippy -p zeroclaw-channels --features channel-discord --lib --no-deps -- -D warnings
    Checking tokio-socks v0.5.2
    Checking idna_adapter v1.2.1
    Checking fluent v0.16.1
    Checking chacha20poly1305 v0.10.1
    Checking hyper-util v0.1.20
    Checking tower-http v0.6.8
    Checking idna v1.1.0
    Checking tokio-stream v0.1.18
    Checking zeroclaw-api v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-api)
    Checking rustls-webpki v0.102.8
    Checking rusqlite v0.37.0
    Checking url v2.5.8
    Checking zeroclaw-infra v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-infra)
   Compiling aws-lc-rs v1.16.3
   Compiling rustls v0.23.39
    Checking rustls-webpki v0.103.13
    Checking tokio-rustls v0.26.4
    Checking tungstenite v0.29.0
    Checking hyper-rustls v0.27.9
    Checking rumqttc v0.25.1
    Checking tokio-tungstenite v0.29.0
    Checking reqwest v0.12.28
    Checking axum v0.8.9
    Checking zeroclaw-config v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-config)
    Checking zeroclaw-memory v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-memory)
    Checking zeroclaw-providers v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-providers)
    Checking zeroclaw-tools v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-tools)
    Checking zeroclaw-runtime v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-runtime)
    Checking zeroclaw-channels v0.7.5 (/home/y00577373/zeroclaw/crates/zeroclaw-channels)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 45.28s
```

- **Commands run:** the three above; all pass with zero warnings.
- **Beyond CI — what I manually verified:** the channel-specific parity tests assert the new path produces the same allow/deny verdict as the old predicate across every truth-table case (wildcard, empty list, specific match, unknown, case sensitivity, substring rejection).
- **Skipped:** `cargo fmt --check` and `cargo test --workspace` are not in the test plan because the relevant invariants are channel-local; the workspace-level checks belong to the merge gate.

## Security & Privacy Impact

- New permissions, capabilities, or file system access scope? **No.**
- New external network calls? **No.**
- Secrets / tokens / credentials handling changed? **No.**
- PII, real identities, or personal data in diff, tests, fixtures, or docs? **No** — test fixtures use synthetic IDs.

This is an access-control refactor: it changes *where* the allowlist check is implemented but not *what* it enforces. Per-channel parity tests assert the new path produces the same verdict as the old predicate across every truth-table case.

## Compatibility

- Backward compatible? **Yes** at the channel config surface — the `allowed_users` config field is still consumed; its value is passed to `AllowlistAspect::new(...)` at channel construction.
- Config / env / CLI surface changed? **No.** Existing operator config files continue to work unchanged.
- Upgrade steps for existing users: none required.

## Rollback (risk: medium)

- **Fast rollback command/path:** `git revert <merge-sha>` reverts this single-channel migration. The chain is designed so each PR is revertable on its own; the only post-revert constraint is that the chain tip (#6799) and any PRs further up the chain that depend on this commit would need to be rebased.
- **Feature flags or config toggles:** None. The aspect is statically linked.
- **Observable failure symptoms:**
  - Allowlist-deny false positives: legitimate `discord` users denied at inbound. Grep logs for `is_allowed → false` followed by message drop in the `discord` channel-level traces.
  - Allowlist-bypass false negatives: unauthorized users let through. Watch for unexpected senders appearing in `chat_history` rows for `discord` if the deployment had a non-wildcard allowlist.
  - Metric to watch: `discord` message-drop rate (delta from steady-state in the hour after deploy).

## Note on the clone warning

The reviewer correctly flagged that `AllowlistAspect::clone` zeros out the `normalizer` and `matcher` fields. The only clone site (`with_allowlist()` for the reaction-worker secondary constructor) uses plain `AllowlistAspect::new()` with no normalizer or matcher closures, so the current clone path is safe. Follow-up filed against `aspect-std` to either (a) make `Clone` carry the closures via `Arc<dyn Fn …>` or (b) emit a compile-time warning when `clone()` is called on an aspect with non-`None` closures. Until then the safe pattern is to share the parent `Arc` rather than clone the aspect; this is the pattern already in use.

### CODE REVIEW

**Comment by Audacity88** [comment]
CI is currently blocked on the new `aspect-std` git dependency. The Windows build cannot fetch `https://github.com/yijunyu/aspect-rs?rev=239ba1d` because that repo contains a path invalid on Windows: `docs/book/src/ch04-core-concepts/*.md`. The Security job is also failing in `cargo-deny` sources, likely from the new git source. Could you switch this to a Windows-checkout-safe / cargo-deny-accepted dependency source before review continues?

**Comment by Audacity88** [comment]
@yijunyu Thanks for updating the dependency source and PR body here. Since this is part of the same AllowlistAspect stack reviewed on #6786–#6800 / #6799, can you include #6638 in the same consolidation plan rather than continuing it as a standalone PR? This branch is also currently conflicting, so I think the cleanest next step is to close/supersede the individual channel PRs and present the consolidated, accurately scoped PR for review.

File: crates/zeroclaw-channels/src/discord.rs [hunk 1]
```diff
@@ -21,7 +22,11 @@ use zeroclaw_api::media::MediaAttachment;
 pub struct DiscordChannel {
     bot_token: String,
     guild_id: Option<String>,
-    allowed_users: Vec<String>,
+    /// Who is allowed to talk to this bot. Wildcard `"*"` admits everyone;
+    /// an empty list denies everyone. Backed by `aspect-std`'s
+    /// `AllowlistAspect` so the check is centralized and reusable across
+    /// channels rather than re-implemented per-module.
+    allowlist: AllowlistAspect,
```
Comment (singlerider) [inline crates/zeroclaw-channels/src/discord.rs:29]: 🔴 **Field no longer exists on `master`.**

The `allowed_users: Vec<String>` field this PR replaces was removed by #6398. Current master's `DiscordChannel` carries a `peer_resolver: Arc<dyn Fn() -> Vec<String> + Send + Sync>` instead (line 37–38), and the allowlist check is delegated to the in-tree `crate::allowlist::is_user_allowed(&peers, user_id, Match::Sensitive)` (discord.rs:236–239).

The `peer_resolver` pattern is deliberate per AGENTS.md's "single source of truth — no cache" rule: the resolver is called at message-time so runtime allowlist changes propagate without restart. Switching to `AllowlistAspect::new(allowed_users)` stored on the struct re-introduces the snapshot pattern this was moved away from.

The constructor signature this PR modifies (`new(bot_token, guild_id, allowed_users, ...)`) is also gone — master's signature is `new(bot_token, guild_ids, alias, peer_resolver, listen_to_bots, mention_only)`. No textual rebase will produce a compiling diff against current master without rethinking the migration model.

File: crates/zeroclaw-channels/Cargo.toml [hunk 2]
```diff
@@ -14,6 +14,10 @@ zeroclaw-memory.workspace = true
 zeroclaw-providers.workspace = true
 zeroclaw-runtime.workspace = true
 zeroclaw-tools.workspace = true
+# AOP support: AllowlistAspect provides the per-channel "who is allowed to
+# talk to the bot" check, replacing 20+ near-identical hand-rolled
+# `fn is_user_allowed` implementations across channel modules.
+aspect-std = "0.1.1"
```
Comment (singlerider) [inline crates/zeroclaw-channels/Cargo.toml:20]: 🔴 **Net-add of external dep for in-tree capability.**

Master already has the centralised allowlist abstraction this PR is reaching for: `crates/zeroclaw-channels/src/allowlist.rs` defines `pub fn is_user_allowed(allowed: &[String], user: &str, mode: Match) -> bool` with explicit `Match::Sensitive` / `Match::CaseInsensitive` variants. Every channel calls it (Discord, Slack, IRC, Telegram, Matrix, etc.) — that's exactly the problem the 24-PR `aspect-std` chain exists to solve.

Reading `crate::allowlist`'s top doc-comment vs the PR body's archetype taxonomy, they're trying to express the same thing. The in-tree solution covers per-channel case-sensitivity policy (`Match::Sensitive` for Discord/Slack/Telegram snowflake-shaped IDs, `Match::CaseInsensitive` for IRC nicks per RFC 2812 and Matrix MXIDs); the `aspect-std::AllowlistAspect` shape used in this PR doesn't carry that policy — see the body discussion.

Adding `aspect-std` here is a net-add of an external dep for a capability that exists in-tree, in a more complete form. The chain-wide call needs to be made before per-channel PRs continue: is `aspect-std` actually adding something `crate::allowlist` doesn't? If yes, the chain should be one consolidation PR that replaces both `crate::allowlist` and `peer_resolver` together; if no, this chain should be closed as obviated by #6398's in-tree work.

File: crates/zeroclaw-channels/src/discord.rs [hunk 3]
```diff
@@ -2156,16 +2174,57 @@ mod tests {
...
+
+    /// Parity test for the AllowlistAspect migration. Asserts that the
+    /// post-migration `allowlist.is_allowed(...)` returns byte-identical
+    /// results to the pre-migration shape
+    /// `allowed.iter().any(|u| u == "*" || u == id)` for a fixed truth table
+    /// covering the wildcard, empty-list, case-sensitivity, and substring
+    /// edge cases. If any reviewer worries the aspect drifted from the
+    /// original Boolean semantics, this test is the regression guard.
+    #[test]
+    fn allowlist_parity_with_pre_aspect_shape() {
```
Comment (singlerider) [inline crates/zeroclaw-channels/src/discord.rs:2197]: 🟢 **This parity test is exactly the right defensive piece.**

A truth table covering wildcard / empty / exact-match / substring-rejection / case-sensitivity, asserting the migrated check is byte-identical to the pre-migration shape — that's the regression guard any allowlist refactor needs. The pattern would be valuable in `crate::allowlist`'s own test module regardless of whether the wider migration goes ahead. If this PR ends up closed for the structural reasons noted above, porting this test (rewritten against `crate::allowlist::is_user_allowed`) as a tiny standalone PR would still be net-positive.

**Review by singlerider — CHANGES_REQUESTED** [review]
I read the current head dc153c2 (May 13), the full diff, both of @Audacity88's prior comments (the Windows-build / `cargo-deny` blocker on May 13 and the consolidation request on May 22), the cross-referenced #6786–#6800 / #6799 stack, and the linked discussion of the migration archetypes. I also cross-checked the proposed change against current `master` (e2fd03223).

The PR is clean, well-scoped, and includes a thoughtful parity test. The supply-chain blocker @Audacity88 raised is genuinely resolved by the `aspect-std 0.1.1` publication. The friction is what's happened to `master` underneath this change — and the answer is: master already has the centralised abstraction this PR is reaching for.

---

### 🟢 What looks good

- **Single-channel scope.** Unlike the earlier multi-channel scope-deception pattern in this series (see my CHANGES_REQUESTED on #6778 and #6793), this PR touches `discord.rs` only. Diff is +110/-29 across `Cargo.lock`, `Cargo.toml`, `discord.rs` — that's it. Good shape.
- **Parity test (`allowlist_parity_with_pre_aspect_shape`).** This is exactly the right defensive piece for an allowlist refactor: a truth table that exercises wildcard / empty-list / exact-match / substring-rejection / case-sensitivity, asserting the migrated check returns byte-identical results to the pre-migration shape. If the aspect ever drifts from the original Boolean semantics this test catches it. The other PRs in this series should adopt the same shape.
- **Supply-chain blocker resolved.** `aspect-std = "0.1.1"` is on crates.io as of 2026-05-24, so the dep is registry-sourced, immutable, visible to `cargo audit` and `cargo deny`. The original Windows-checkout-unsafe git source is gone. @Audacity88's May 13 blocker is fully addressed.

---

### 🔴 Blocking — master already has the centralised abstraction this PR re-introduces

This is the big one and it changes the framing of the whole 24-PR chain.

After #6398 (`feat!: multi-agent runtime and schema V3`, merged May 20), `discord.rs` no longer carries an `allowed_users: Vec<String>` field. The shape is now:

```rust
// crates/zeroclaw-channels/src/discord.rs:37–38 (current master)
peer_resolver: Arc<dyn Fn() -> Vec<String> + Send + Sync>,
```

And the `is_user_allowed` predicate is now:

```rust
// crates/zeroclaw-channels/src/discord.rs:236–239
fn is_user_allowed(&self, user_id: &str) -> bool {
    let peers = (self.peer_resolver)();
    crate::allowlist::is_user_allowed(&peers, user_id, crate::allowlist::Match::Sensitive)
}
```

`crate::allowlist` (in `crates/zeroclaw-channels/src/allowlist.rs`) is the in-tree single source of truth for the allowlist semantics — wildcard, empty-list, case-sensitive vs case-insensitive — that this PR aims to centralise via `aspect-std`. Reading its top doc-comment is striking:

> Each channel (Slack, Discord, IRC, Telegram, Matrix, …) carries an `allowed_users: Vec<String>` allowlist with the same semantics:
> - `["*"]` (or any list containing `"*"`) means "anyone".
> - Empty list means "deny everyone"
> - Otherwise, exact match against the user's identifier wins.
> IRC nicks are case-insensitive per RFC 2812; Matrix MXIDs are also case-insensitive. Most other channels (Slack user IDs, Discord snowflakes, Telegram usernames) are case-sensitive. The `Match::Sensitive` / `Match::CaseInsensitive` selector encodes that per-channel choice without growing a parallel impl.

That's the exact problem this PR series exists to solve. Master already solved it.

Two consequences:

1. **Adopting `aspect-std::AllowlistAspect` is a net-add of an external dep for capability that exists in-tree.** The principal benefit of `aspect-std` — centralising a check across modules — is already delivered by `crate::allowlist::is_user_allowed`. The PR's archetype-A semantics (wildcard, byte-equality) are a *subset* of what `crate::allowlist` already covers (Sensitive + CaseInsensitive variants).
2. **The migration also drops the runtime-resolver pattern.** Master stores `peer_resolver: Arc<dyn Fn() -> Vec<String> + Send + Sync>` rather than a snapshot `Vec<String>`. The resolver is called at message-time so allowlist changes propagate without restart (AGENTS.md's "single source of truth — no cache" rule). Switching to `AllowlistAspect::new(allowed_users)` stored on the struct re-introduces the snapshot pattern this was deliberately moved away from.

So the rebase against current `master` isn't just textual — it's "the codebase moved through and past where this PR was trying to take it." The aspect-std crate is solving a problem that no longer exists in-tree.

---

### 🔴 Blocking — case-sensitivity regression for the wider chain

The PR body classifies discord as "Archetype A — flat-list + `*` byte-equality" (correct for Discord snowflakes — case-sensitive). But the wider 24-PR chain has channels that are deliberately case-insensitive: IRC (RFC 2812 nicks), Matrix (MXIDs). Those channels currently pass `Match::CaseInsensitive` to `crate::allowlist::is_user_allowed`.

`aspect-std::AllowlistAspect` — at least as used here — has no case-sensitivity selector visible in the diff. If the same `AllowlistAspect::new(allowed_users)` shape gets applied to IRC and Matrix in the wider chain, it silently downgrades case-insensitive matching to case-sensitive matching. RFC 2812 nick `alice` would no longer match `Alice`. That's an authorization regression dressed up as a refactor.

This isn't blocking on the discord PR specifically — Discord is genuinely case-sensitive. But the chain-wide goal of "use AllowlistAspect everywhere" runs into this for IRC and Matrix specifically, and the consolidation @Audacity88 asked for on May 22 should address how the aspect carries per-channel case-sensitivity (or whether it should at all). Worth knowing now rather than discovering when the IRC/Matrix archetype PRs land.

---

### 🟡 Warning — branch CONFLICTING and the consolidation ask is unaddressed

@Audacity88's May 22 ask was to fold #6638 into the chain consolidation alongside #6786–#6800 / #6799 rather than continuing it as a standalone. Six days later the branch is still CONFLICTING and there's no consolidated PR. Whatever the final shape is — consolidated or per-channel — it needs to be against current `master`, which means against the in-tree `crate::allowlist` reality.

---

### Verdict

The instinct behind this work — centralise allowlist logic that was duplicated per-channel — was the right one in April when the chain was opened. Master got there first by a different route in May (in-tree `crate::allowlist` + `peer_resolver`). The right move now is to step back, evaluate whether `aspect-std` offers anything the in-tree solution doesn't (one candidate: per-channel case-sensitivity policy as a typed field on the aspect, instead of a `Match` enum argument at every call site), and re-scope the chain accordingly.

If the answer is "yes, `aspect-std` is the better long-term path" — then the right unit of work is a single chain consolidation PR that replaces `crate::allowlist` and `peer_resolver` together, with a clear before/after on case-sensitivity coverage and a rationale for why the external crate beats the in-tree one. Per-channel migration PRs against the *current* master can't compile cleanly because the field they're replacing (`allowed_users: Vec<String>`) no longer exists.

If the answer is "no, the in-tree solution is fine" — then closing this PR (and the chain) as obviated by #6398's in-tree centralisation is the honest call. The parity-test pattern alone is valuable enough that it'd be worth porting into `crate::allowlist`'s own test module on a one-line PR.

Either way, I don't think this PR as it stands has a path to merge against current master. The friction isn't reviewer push-back; it's that the code moved.

Thanks for the careful work on the parity test and the supply-chain resolution. The instinct was right, even if the timing got bypassed.

**Comment by yijunyu** [comment]
Closing. This channel already routes its allowlist check through the in-tree `crate::allowlist::is_user_allowed(&peers, id, Match::…)` that landed in #6398 — the same consolidation this PR's `AllowlistAspect` migration was proposing, done in-tree with no external dependency and resolving the peers fresh from config per call (single source of truth).

Re-applying the aspect on top of that would replace working in-tree code with an external crate for no behavioral gain, so this PR is superseded.

The aspect still lands where `crate::allowlist`'s flat-list matching structurally can't reach — the channels whose identity matching needs normalization the `Match` enum can't express. Those migrations are open as #7017 (whatsapp_web, E.164 phone) and #7018 (email + gmail_push, domain-class), using a stateless `aspect_std::is_allowed_in` (crates.io `aspect-std 0.1.2`) that holds no state and so honors the single-source-of-truth rule.

Thanks to the reviewers who looked at this stack.
